Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean

```diff
@@ -532,117 +526,127 @@ section ModelsWithJ
-  have := neZero_iff.2 <| two_or_three_ne_zero.neg_resolve_left h2
-  ofJ_0_of_three_ne_zero
-
-lemma ofJ_1728_of_three_eq_zero (h3 : (3 : F) = 0) :
-    ofJ 1728 = @ofJ1728 _ _ (invertibleOfNonzero <| two_or_three_ne_zero.neg_resolve_right h3) := by
-  rw [ofJ, dif_pos <| by rw [show (1728 : F) = 3 * 576 by norm_num1, h3, zero_mul], dif_pos h3]
-
-lemma ofJ_1728_of_two_ne_zero [h2 : NeZero (2 : F)] :
-    ofJ 1728 = @ofJ1728 _ _ (invertibleOfNonzero h2.out) := by
+with j-invariant equal to j (see `WeierstrassCurve.ofJ_j`).
+Its coefficients are given explicitly (see `WeierstrassCurve.ofJ0`, `WeierstrassCurve.ofJ1728`
+and `WeierstrassCurve.ofJNe0Or1728`). -/
+def ofJ : WeierstrassCurve F :=
+  if j = 0 then if (3 : F) = 0 then ofJ1728 F else ofJ0 F
+  else if j = 1728 then ofJ1728 F else ofJNe0Or1728 j
+
+lemma ofJ_0_of_three_ne_zero [h3 : NeZero (3 : F)] : ofJ 0 = ofJ0 F := by
+  rw [ofJ, if_pos rfl, if_neg h3.out]
+
+lemma ofJ_0_of_three_eq_zero (h3 : (3 : F) = 0) : ofJ 0 = ofJ1728 F := by
+  rw [ofJ, if_pos rfl, if_pos h3]
+
+lemma ofJ_0_of_two_eq_zero (h2 : (2 : F) = 0) : ofJ 0 = ofJ0 F := by
+  rw [ofJ, if_pos rfl, if_neg ((show (3 : F) = 1 by linear_combination h2) ▸ one_ne_zero)]
+
+lemma ofJ_1728_of_three_eq_zero (h3 : (3 : F) = 0) : ofJ 1728 = ofJ1728 F := by
+  rw [ofJ, if_pos (show (1728 : F) = 0 by linear_combination 576 * h3), if_pos h3]
+
+lemma ofJ_1728_of_two_ne_zero [h2 : NeZero (2 : F)] : ofJ 1728 = ofJ1728 F := by
   by_cases h3 : (3 : F) = 0
-  · exact ofJ_1728_of_three_eq_zero h3
-  · have h : (1728 : F) ≠ 0 := fun h => or_iff_not_and_not.mp
-      (mul_eq_zero.mp <| by rwa [show 2 ^ 6 * 3 ^ 3 = (1728 : F) by norm_num1])
-      ⟨pow_ne_zero 6 h2.out, pow_ne_zero 3 h3⟩
-    rw [ofJ, dif_neg h, dif_pos rfl]
+  · rw [ofJ, if_pos (show (1728 : F) = 0 by linear_combination 576 * h3), if_pos h3]
+  · rw [ofJ, if_neg ((show 2 ^ 6 * 3 ^ 3 = (1728 : F) by norm_num1) ▸ mul_ne_zero
+      (pow_ne_zero 6 h2.out) (pow_ne_zero 3 h3)), if_pos rfl]
 
-lemma ofJ_1728_of_two_eq_zero (h2 : (2 : F) = 0) :
-    ofJ 1728 = @ofJ0 _ _ (invertibleOfNonzero <| two_or_three_ne_zero.neg_resolve_left h2) := by
-  rw [ofJ, dif_pos <| by rw [show (1728 : F) = 2 * 864 by norm_num1, h2, zero_mul], dif_neg]
+lemma ofJ_1728_of_two_eq_zero (h2 : (2 : F) = 0) : ofJ 1728 = ofJ0 F := by
+  rw [ofJ, if_pos (show (1728 : F) = 0 by linear_combination 864 * h2),
+    if_neg ((show (3 : F) = 1 by linear_combination h2) ▸ one_ne_zero)]
 
-lemma ofJ_ne_0_ne_1728 (h0 : j ≠ 0) (h1728 : j ≠ 1728) : ofJ j =
-    @ofJ' _ _ j (invertibleOfNonzero h0) (invertibleOfNonzero <| sub_ne_zero_of_ne h1728) := by
-  rw [ofJ, dif_neg h0, dif_neg h1728]
+lemma ofJ_ne_0_ne_1728 (h0 : j ≠ 0) (h1728 : j ≠ 1728) : ofJ j = ofJNe0Or1728 j := by
+  rw [ofJ, if_neg h0, if_neg h1728]
+
+instance : (ofJ j).IsElliptic := by
+  by_cases h0 : j = 0
+  · by_cases h3 : (3 : F) = 0
+    · haveI := Fact.mk (isUnit_of_mul_eq_one (2 : F) 2 (by linear_combination h3))
```

Comment:
All the `haveI` can be replaced by `have`. Once we replace `Fact (IsUnit ...)` with `IsUnit ...`, and once we decide on whether to remove `NeZero`, most of these should be a lot cleaner to prove.

Reply:
I just get used to write `haveI` for instances instead of `have`. Anyways, I'll change them.